src/coreclr/src/jit/lclvars.cpp

```diff
@@ -3640,6 +3646,53 @@ var_types LclVarDsc::lvaArgType()
     return type;
 }
 
+//----------------------------------------------------------------------------------------------
+// CanBeReplacedWithItsField: check if a whole struct reference could be replaced by a field.
+//
+// Arguments:
+//    comp - the compiler instance;
+//
+// Return Value:
+//    true if that can be replaced, false otherwise.
+//
+// Notes:
+//    The replacement can be made only for independently promoted structs
+//    with 1 field without holes.
+//
+bool LclVarDsc::CanBeReplacedWithItsField(Compiler* comp) const
+{
+    if (!lvPromoted)
+    {
+        return false;
+    }
+
+    if (comp->lvaGetPromotionType(this) != Compiler::PROMOTION_TYPE_INDEPENDENT)
+    {
+        return false;
+    }
+    if (lvFieldCnt != 1)
+    {
+        return false;
+    }
+    if (lvContainsHoles)
+    {
+        return false;
+    }
+
+#if defined(FEATURE_SIMD)
+    // If we return `struct A { SIMD16 a; }` we split the struct into several fields.
+    // In order to do that we have to have its field in memory. Right now lowering cannot
+    // handle RETURN struct(multiple registers)->SIMD16(one register), but it can be improved.
+    LclVarDsc* fieldDsc = comp->lvaGetDesc(lvFieldLclStart);
+    if (fieldDsc->TypeGet() == TYP_SIMD12 || fieldDsc->TypeGet() == TYP_SIMD16)
```

Comment:
Changed, but not sure that I understood it right, please check.

Reply:
I have tried to fix that, but now sure that I understood that correctly, please check.

Reply:
I think the issue isn't whether the parent struct is an opaque SIMD type, but whether the contained struct is. We do promote a case like `struct A { SIMD16 a; }` so that the SIMD16 can be in a register. If it is an opaque struct, we will pass it in a single register on Arm64 - and also on X64/linux once #9578 is fixed. So I think it's the field handle that needs to be checked. I guess we are still inhibiting struct promotion for x64/ux because these are currently handled as "vanilla" structs.

Reply:
Fixed.